tests/storage/test_event_chain.py

```diff
@@ -0,0 +1,494 @@
+                        "type": event.type,
+                        "state_key": event.state_key,
+                    }
+                    for event in events
+                ],
+            )
+
+            # Actually call the function that calculates the auth chain stuff.
+            persist_events_store._persist_event_auth_chain_txn(txn, events)
+
+        self.get_success(
+            persist_events_store.db_pool.runInteraction("_persist", _persist,)
+        )
+
+    def fetch_chains(
+        self, events: List[EventBase]
+    ) -> Tuple[Dict[str, Tuple[int, int]], _LinkMap]:
+
+        # Fetch the map from event ID -> (chain ID, sequence number)
+        rows = self.get_success(
+            self.store.db_pool.simple_select_many_batch(
+                table="event_auth_chains",
+                column="event_id",
+                iterable=[e.event_id for e in events],
+                retcols=("event_id", "chain_id", "sequence_number"),
+                keyvalues={},
+            )
+        )
+
+        chain_map = {
+            row["event_id"]: (row["chain_id"], row["sequence_number"]) for row in rows
+        }
+
+        # Fetch all the links and pass them to the _LinkMap.
+        rows = self.get_success(
+            self.store.db_pool.simple_select_many_batch(
+                table="event_auth_chain_links",
+                column="origin_chain_id",
+                iterable=[chain_id for chain_id, _ in chain_map.values()],
+                retcols=(
+                    "origin_chain_id",
+                    "origin_sequence_number",
+                    "target_chain_id",
+                    "target_sequence_number",
+                ),
+                keyvalues={},
+            )
+        )
+
+        link_map = _LinkMap()
+        for row in rows:
+            added = link_map.add_link(
+                row["origin_chain_id"],
+                row["origin_sequence_number"],
+                row["target_chain_id"],
+                row["target_sequence_number"],
+            )
+
+            # We shouldn't have persisted any redundant links
+            self.assertTrue(added)
+
+        return chain_map, link_map
+
+
+class LinkMapTestCase(unittest.TestCase):
+    def test_simple(self):
+        """Basic tests for the LinkMap.
+        """
+        link_map = _LinkMap()
+
+        link_map.add_link(1, 1, 2, 1, new=False)
+        self.assertYieldsUnordered(link_map.get_links_between(1, 2), [(1, 1)])
+        self.assertYieldsUnordered(link_map.get_links_from(1, 1), [(2, 1)])
+        self.assertYieldsUnordered(link_map.get_additions(), [])
+        self.assertTrue(link_map.exists_path_from(1, 5, 2, 1))
+        self.assertFalse(link_map.exists_path_from(1, 5, 2, 2))
+        self.assertTrue(link_map.exists_path_from(1, 5, 1, 1))
+        self.assertFalse(link_map.exists_path_from(1, 1, 1, 5))
+
+        # Attempting to add a redundant link is ignored.
+        link_map.add_link(1, 4, 2, 1)
```

Comment:
I'm guessing we want to ensure this doesn't get added?

```suggestion
        self.assertFalse(link_map.add_link(1, 4, 2, 1))
```

Reply:
Oh yes, I added the `add_link` return value after writing these tests